recovery/elliptics_recovery/types/dc.py

```diff
@@ -222,23 +227,71 @@ def contains(addresses_with_backends, address, backend_id):
     return address_range
 
 
-def final_merge(ctx, results):
-    import shutil
-    ctx.stats.timer('main', 'final_merge')
-    log.info("final merge")
-
-    ctx.merged_filename = os.path.join(ctx.tmp_dir, 'merged_result')
-    dump_filename = os.path.join(ctx.tmp_dir, 'dump')
-    with open(ctx.merged_filename, 'wb') as mf:
-        with open(dump_filename, 'wb') as df:
-            for res in (r for r in results if r):
-                shutil.copyfileobj(open(res[0], 'rb'), mf)
-                os.remove(res[0])
-                shutil.copyfileobj(open(res[1], 'rb'), df)
-                os.remove(res[1])
-
-    log.debug("Merged_filename: %s, address: %s, groups: %s, tmp_dir: %s",
-              ctx.merged_filename, ctx.address, ctx.groups, ctx.tmp_dir)
+def dump_key(ctx, rest_file, key, key_infos, newest_key_group):
+    if key_infos[0].group_id != newest_key_group:
+        index = [k.group_id for k in key_infos].index(newest_key_group)
+        tmp = list(key_infos)
+        tmp[0], tmp[index] = tmp[index], tmp[0]
+        key_infos = tuple(tmp)
+
+    key_data = (key, key_infos)
+    key_info = key_infos[0]
+    is_chunked = key_info.size > ctx.chunk_size
+    same_ts = lambda lhs, rhs: lhs.timestamp == rhs.timestamp
+    same_uncommitted = [info for info in key_infos
+                        if info.flags & elliptics.record_flags.uncommitted and same_ts(info, key_info)]
+    is_all_uncommitted = same_uncommitted == key_infos and key_info.timestamp < ctx.prepare_timeout
+
+    log.debug("Dumping key: {0}, group: {1}".format(key_data[0], newest_key_group))
+    if is_chunked or is_all_uncommitted:
+        dump_key_data(key_data, rest_file)
+    else:
+        if newest_key_group not in ctx.bucket_files:
+            filename = os.path.join(ctx.tmp_dir, 'bucket_%d' % (newest_key_group))
+            ctx.bucket_files[newest_key_group] = open(filename, 'wb+')
+        dump_key_data(key_data, ctx.bucket_files[newest_key_group])
+
+
+'''
+This function distributes keys among multiple files (buckets).
+One bucket is 'rest_keys' and other buckets are 'bucket_xx', where xx == group_id.
+'bucket_xx' contains newest keys that should be recovered from group xx to other groups
+via server_send. If a key could not be recovered with server_send it is placed to 'rest_keys'.
+
+Also this function prepares bucket_order array. The array contains group_id's sorted by amount
+of keys in appropriate bucket. This array will be used by server_send recovery process.
+'''
+def fill_buckets(ctx, results):
+    newest_key_stats = dict()
+    for _, _, range_stats in (r for r in results if r):
```

Comment:
As I see, wherever `results` are used they are filtered via `(r for r in results if r)`. It can be done ones [here] (https://github.com/reverbrain/elliptics/pull/668/files#diff-302f790c9667058c7ff408b010833a2fR328).